test/functional/feature_securethebag.py

```diff
@@ -0,0 +1,184 @@
+    outputs_tree = [[CTxOut()]*(2**i) for i in range(depth)] + [leaf_nodes]
+    for d in range(1, depth+2):
+        idxs =zip(range(0, len(outputs_tree[-d]),2), range(1, len(outputs_tree[-d]), 2))
+        for (idx, (a,b)) in enumerate([(outputs_tree[-d][i], outputs_tree[-d][j]) for (i,j) in idxs]):
+            s = CScript(bytes([OP_SECURETHEBAG, 0x20]) + bag_for_outputs([a,b]))
+            a = sum(o.nValue for o in [a,b])
+            t = CTxOut(a+1000, s)
+            outputs_tree[-d-1][idx] = t
+    return outputs_tree
+
+def create_transaction_to_script(node, txid, script, *, amount):
+    """ Return signed transaction spending the first output of the
+        input txid. Note that the node must be able to sign for the
+        output that is being spent, and the node must not be running
+        multiple wallets.
+    """
+    random_address = script_to_p2sh(CScript())
+    rawtx = node.createrawtransaction(inputs=[{"txid": txid, "vout": 0}], outputs={random_address: amount})
+    tx = CTransaction()
+    tx.deserialize(BytesIO(hex_str_to_bytes(rawtx)))
+    # Replace with our script
+    tx.vout[0].scriptPubKey = script
+    # Sign
+    signresult = node.signrawtransactionwithwallet(tx.serialize().hex())
+    assert_equal(signresult["complete"], True)
+    raw_tx =  signresult['hex']
+    tx = CTransaction()
+    tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx)))
+    return tx
+
+class SecureTheBagTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-whitelist=127.0.0.1', '-par=1']]  # Use only one script thread to get the exact reject reason for testing
+        self.setup_clean_chain = True
+        self.rpc_timeout = 120
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+    def get_block(self, txs):
+        self.tip    = self.nodes[0].getbestblockhash()
+        self.height = self.nodes[0].getblockcount()
+        block = create_block(int(self.tip, 16), create_coinbase(self.height))
+        block.vtx.extend(txs)
+        add_witness_commitment(block)
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        return block.serialize(True).hex(), block.hash
+    def add_block(self, txs):
+        block, h = self.get_block(txs)
+        self.nodes[0].submitblock(block)
+        assert_equal(self.nodes[0].getbestblockhash(), h)
+        return h
+    def fail_block(self, txs, cause = SECURETHEBAG_ERROR):
+        block, h = self.get_block(txs)
+        assert_equal(self.nodes[0].submitblock(block), cause)
+        assert_equal(self.nodes[0].getbestblockhash(), self.tip)
+
+    def run_test(self):
+        self.nodes[0].add_p2p_connection(P2PInterface())
+
+        BLOCKS = 103
+        self.log.info("Mining %d blocks for mature coinbases", BLOCKS)
+        self.coinbase_txids = [self.nodes[0].getblock(b)['tx'][0] for b in self.nodes[0].generate(BLOCKS)]
```

Comment:
can clean this up by making it an `iter` and then calling `next(self.coinbase_txids)` instead of hardcoding indices when accessing

Reply:
but does this actually need to be an instance attribute? why not just a local variable?

Reply:
So that doesn't quite work I think because we need to generate all the blocks now for maturity, and an iter would be lazy

Reply:
you're thinking `iter(a for a in b)` which would be lazy instead of `iter([a for a in b])` which would provide an iterator over the already-evaluated list

Reply:
ipython output example:
```
In [1]: iter([print(i) for i in range(5)])                                     
0
1
2
3
4
Out[1]: <list_iterator at 0x7f172c2cd908>

In [2]: iter(print(i) for i in range(5))                                       
Out[2]: <generator object <genexpr> at 0x7f172c2d5888>
```